// htmltree-stringify/index.js
'use strict';

// modify from https://github.com/defunctzombie/node-taters/blob/master/lib/render.js

var Output = require('./output');

module.exports = function stringify(doc) {
  var out = new Output();
  var write = out.write.bind(out);

  doctype(write, doc);
  doc.root.forEach(render.bind(null, write));

  return out.data;
};

function doctype(write, doc) {
  if (doc.doctype) {
    write('<!doctype' + doc.doctype + '>');
  }
}

function render(write, element) {
  text(write, element)
    || comment(write, element)
    || tag(write, element);
}

function text(write, element) {
  if (element.type !== 'text') {
    return;
  }
  return write(element.data);
}

function comment(write, element) {
  if (element.type !== 'comment') {
    return;
  }
  return write('<!--' + element.data + '-->');
}

function tag(write, element) {
  if (element.type !== 'tag') {
    return;
  }

  element.attributes = element.attribs || element.attributes;

  write('<' + element.name);

  if (element.attributes) {
    Object
      .keys(element.attributes)
      .forEach(function (key) {
        write(' ' + key + '="' + element.attributes[key] + '"');
      });
  }

  if (element.void) {
    return write('/>');
  }

  write('>');

  if (element.data) {
    write(element.data);
  }

  element.children.forEach(render.bind(null, write));

  return write('</' + element.name + '>');
}
